src/LockTable.h

```diff
@@ -0,0 +1,187 @@
+ * are added.
+ *
+ * For best performance, the number of buckets should be set large enough so
+ * that overflow cache lines are almost never needed but small enough that the
+ * entire structure might fit in CPU cache.
+ */
+class LockTable {
+  PUBLIC:
+    explicit LockTable(uint64_t numBuckets);
+    virtual ~LockTable();
+
+    void acquireLock(uint64_t lockId);
+    bool isLockAcquired(uint64_t lockId);
+    void releaseLock(uint64_t lockId);
+    bool tryAcquireLock(uint64_t lockId);
+
+  PRIVATE:
+    /**
+     * A lock table entry.
+     *
+     * Lock table entries live in \link CacheLine CacheLines\endlink.  A lock
+     * table entry holds the lockId of an acquired lock.
+     */
+    typedef uint64_t Entry;
+    static_assert(sizeof(Entry) == 8, "LockTable::Entry is not 8 bytes");
+
+    /**
+     * The number of bytes per cache line in this machine.
+     */
+    static const uint32_t BYTES_PER_CACHE_LINE = 64;
+
+    /**
+     * The number of LockTable Entry objects in a CacheLine.
+     */
+    static const uint32_t ENTRIES_PER_CACHE_LINE = (BYTES_PER_CACHE_LINE /
+                                                    sizeof(Entry)) - 1;
+    static_assert(BYTES_PER_CACHE_LINE % sizeof(Entry) == 0,
+                  "BYTES_PER_CACHE_LINE not a multiple of sizeof(Entry)");
+
+    /**
+     * A linked list of cache lines composes a bucket within the LockTable.
+     *
+     * Each cache line is composed of several lock table entries and a pointer
+     * to the next overflow cache line (if any).
+     *
+     * A CacheLine is meant to fit on a single L2 cache line on the CPU.
+     * Different processors may require tweaking ENTRIES_PER_CACHE_LINE to
+     * achieve this.
+     */
+    struct CacheLine {
+        /// Array of lock table entries (acquired lockIds).
+        Entry entries[ENTRIES_PER_CACHE_LINE];
+        /// Pointer to next cache line in bucket; NULL if last in chain.
+        CacheLine* next;
+    };
+    static_assert(sizeof(CacheLine) == BYTES_PER_CACHE_LINE,
+                  "sizeof(CacheLine) does not match BYTES_PER_CACHE_LINE");
+
+    /**
+     * The BucketLock is a spin-lock style lock used to prevent concurrent
+     * access to individual LockTable buckets.
+     *
+     * Bucket locks live in memory of the first lock table Entry of a bucket's
+     * first cache line.  As such, bucket locks must fit in a lock table entry's
+     * memory footprint.
+     */
+    struct BucketLock {
+        /**
+         * Blocks until a lock can be obtained.
+         */
+        void lock()
+        {
+            while (mutex.exchange(1) != 0)
+                continue;
+
+            Fence::enter();
+        }
+
+        /**
+         * Attempts to acquire the lock for without blocking.
+         *
+         * \return
+         *      TRUE if the lock was acquired, FALSE otherwise.
+         */
+        bool try_lock()
+        {
+            uint64_t old = mutex.exchange(1);
+            if (old == 0) {
```

Comment:
Nit: can't you get rid of the variable "old" ? This would make the structure more similar to that of the lock method.